allennlp/data/dataset_readers/dataset_utils/span_utils.py

```diff
@@ -401,42 +403,139 @@ def extract_bmes_tag_label(text):
+            ('B', 'E'): (append_and_close_span, close_then_create_and_close_span),  # (O, X)
+            ('B', 'S'): (close_then_create_and_close_span, None),                   # (X, X)
+            # Transision from M.
+            ('M', 'B'): (close_then_create_span, None),                             # (X, X)
+            ('M', 'M'): (append_and_keep_span, close_then_create_span),             # (O, X)
+            ('M', 'E'): (append_and_close_span, close_then_create_and_close_span),  # (O, X)
+            ('M', 'S'): (close_then_create_and_close_span, None),                   # (X, X)
+            }
+    # Key (x, y): transition from x to y.
+    # Val func: op for transition from x to y.
+    # We don't need to consider label mismatch issue here.
+    from_es_op = {
+            # Transision from E.
+            ('E', 'B'): create_and_keep_span,   # O
+            ('E', 'M'): create_and_keep_span,   # X
+            ('E', 'E'): create_and_close_span,  # X
+            ('E', 'S'): create_and_close_span,  # O
+            # Transision from S.
+            ('S', 'B'): create_and_keep_span,   # O
+            ('S', 'M'): create_and_keep_span,   # X
+            ('S', 'E'): create_and_close_span,  # X
+            ('S', 'S'): create_and_close_span,  # O
+            }
+
+    spans: List[TypedStringSpan] = []
+    state: Dict[str, Union[str, int, bool]] = {}
+
+    for index, raw_tag in enumerate(tag_sequence):
+        cur_bmes_tag, cur_label = extract_bmes_tag_label(raw_tag)
+        # Only BMES tags are accepted.
+        if cur_bmes_tag not in ('B', 'M', 'E', 'S'):
+            raise InvalidTagSequence(tag_sequence)
 
-            if index >= len(tag_sequence):
-                invalid = True
-            if not invalid:
-                spans.append((start_label, (start_index, index)))
+        kwargs = {
+                'bmes_tag': cur_bmes_tag,
+                'label': cur_label,
+                'index': index,
+                }
 
-        elif start_bmes_tag == 'S':
-            spans.append((start_label, (start_index, start_index)))
+        # The first element, create a span anyway.
+        if index == 0:
+            func = init_op[cur_bmes_tag]
+            func(spans, state, **kwargs)
+            continue
 
-        else:
-            invalid = True
+        last_label = span_label(state)
+        last_bmes_tag = span_last_bmes_tag(state)
+        op_key = (last_bmes_tag, cur_bmes_tag)
 
-        # Move to next span.
-        index += 1
+        if last_bmes_tag in ('B', 'M'):
+            func_match, func_mismatch = from_bm_op[op_key]
+            if func_mismatch and last_label != cur_label:
+                func_mismatch(spans, state, **kwargs)
+            else:
+                func_match(spans, state, **kwargs)
+        else:
+            func = from_es_op[op_key]
+            func(spans, state, **kwargs)
 
-    if invalid:
-        raise InvalidTagSequence(tag_sequence)
```

Comment:
This change in logic seems way more complicated than it needs to be.  I think the new logic that you have is way harder to understand than what you had before, and I don't think we need the complicated state transition logic just to handle this simple transition system.  Can't you just add the tag sequence here?  If you really want to be smart about handling `["M", "E"]` as a single span, you can still do that with small modifications to the previous if/else system of handling the transition system.

My motivation: if there's a bug in this code somewhere down the line, or we decide we need to refactor things, we'll be responsible for maintaining it.  I think the logic that you had before is simple enough that I can maintain it without much work.  This new logic is _much_ too complex for me to want to maintain.  Please keep the tests that you wrote, but make minimal modifications to the old logic to get the tests to pass.

Reply:
> This new logic is much too complex for me to want to maintain.

I totally understand your concern. But I think the code will be complicated somehow.
As an alternative, I wonder if we can support passing an optional `Callable` to `SpanBasedF1Measure` (to override `self. label_encoding`) so that user could provide and maintain the span generator by themselves.

https://github.com/allenai/allennlp/blob/7bf930f48ff8a72611db61301268b058e5cbaf48/allennlp/training/metrics/span_based_f1_measure.py#L38

https://github.com/allenai/allennlp/blob/7bf930f48ff8a72611db61301268b058e5cbaf48/allennlp/training/metrics/span_based_f1_measure.py#L138-L149

Reply:
@DeNeutoy, do you have an opinion on how to handle this?  You've done a lot more with the span logic than I have.  It looks like most of these tags-to-span methods just throw errors on invalid sequences, though the BIO one doesn't.  What should we do?